[CMS-458] Add `version_constraint` input argument

jabxjab opened the pull request.

Commit:
Add "version_constraint" option

Commit:
Add tests

Commit:
Refactor getMajorVersionConstraint() into

Commit:
Move version constraint check from getReleasesFromGithub() into getLatestReleaseFromGithub()

greg-1-anderson commented on src/SelfUpdateCommand.php:
```diff
@@ -198,28 +233,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
             $output->writeln('<info>Successfully updated ' . $programName . '</info>');
-            $this->_exit();
+
+            exit;
```
Methods that call `exit` are untestable. I'd prefer to keep the `_exit()` wrapper even if we aren't trying to run unit tests on this method.

jabxjab replied:
Restored

greg-1-anderson commented on src/SelfUpdateCommand.php:
```diff
@@ -48,6 +68,7 @@ protected function configure()
             ->addOption('stable', NULL, InputOption::VALUE_NONE, 'Use stable releases (default)')
             ->addOption('preview', NULL, InputOption::VALUE_NONE, 'Preview unstable (e.g., alpha, beta, etc.) releases')
             ->addOption('compatible', NULL, InputOption::VALUE_NONE, 'Stay on current major version')
+            ->addOption('version_constraint', NULL, InputOption::VALUE_REQUIRED, 'Apply version constraint')
```
Please make this an optional argument rather than an option (e.g. `self:update ^2` to roll back to v2)

jabxjab replied:
Changed

greg-1-anderson commented on src/SelfUpdateCommand.php:
```diff
@@ -101,38 +127,54 @@ protected function getReleasesFromGithub()
         }
 
-        return [ $version, $url ];
+        return null;
```
Stylistically I prefer having a uniform return type, and leave the current version check outside of this method. Optional.

jabxjab replied:
The current version check moved outside of the method

greg-1-anderson commented on src/SelfUpdateCommand.php:
```diff
@@ -101,38 +127,54 @@ protected function getReleasesFromGithub()
-        $url = null;
-
-        foreach ($releases as $release) {
+    public function getLatestReleaseFromGithub() {
```
Stylistically I prefer passing local variables to using class fields, as it makes methods easier to understand. The best method vis-a-vis testability is one that takes parameters and returns a result, with no side effects. This is a good design aesthetic even when we're not providing unit tests yet. Optional.

jabxjab replied:
TBH, was hesitating about this approach initially, refactored to get rid of class fields.

greg-1-anderson commented on src/SelfUpdateCommand.php:
```diff
@@ -25,6 +25,26 @@ class SelfUpdateCommand extends Command
+    /**
+     * @var bool
+     */
+    protected $isCompatible = false;
```
Naming things is hard: `isCompatible` is hard to understand because there is no comment, takes some reading to see that this is the "compatible" option. `$isCompatibleOptionSelected`? Optional.

jabxjab replied:
Fixed

greg-1-anderson reviewed (approved):
Looks good, just had some style-related notes.

Commit:
Add input argument instead of option

Commit:
Remove option properties
Pass option/arg properties as args
Move out "The latest release matches the current one" check off of getLatestReleaseFromGithub()
Restore _exit()

Commit:
Update tests

greg-1-anderson reviewed (approved):
LGTM
